packages/devtools_app/lib/src/app.dart

```diff
@@ -99,111 +96,122 @@ class DevToolsAppState extends State<DevToolsApp> {
             args,
-          );
-      assert(() {
-        builder = (context) => _AlternateCheckedModeBanner(
-              builder: (context) => routes[path](
-                context,
-                uri.queryParameters,
-                args,
-              ),
-            );
+          ),
+        );
         return true;
       }());
-      return MaterialPageRoute(settings: settings, builder: builder);
+      return MaterialPage(builder: (context) => widget);
     }
 
     // Return a page not found.
-    return MaterialPageRoute(
-      settings: settings,
-      builder: (BuildContext context) {
-        return DevToolsScaffold.withChild(
-          child: CenteredMessage("'$uri' not found."),
-          ideTheme: ideTheme,
-          analyticsProvider: widget.analyticsProvider,
-        );
-      },
+    return MaterialPage(
+      builder: (context) => DevToolsScaffold.withChild(
+        key: const Key('not-found'),
+        child: CenteredMessage("'$page' not found."),
+        ideTheme: ideTheme,
+        analyticsProvider: widget.analyticsProvider,
+      ),
     );
   }
 
-  /// The routes that the app exposes.
-  Map<String, UrlParametersBuilder> get routes {
-    return _routes ??= {
-      homeRoute: (_, params, __) {
-        if (params['uri']?.isNotEmpty ?? false) {
-          final embed = params['embed'] == 'true';
-          final page = params['page'];
-          return Initializer(
-            url: params['uri'],
-            allowConnectionScreenOnDisconnect: !embed,
-            builder: (_) {
-              final tabs = embed && page != null
-                  ? _visibleScreens().where((p) => p.screenId == page).toList()
-                  : _visibleScreens();
-              if (tabs.isEmpty) {
-                return DevToolsScaffold.withChild(
-                  child: CenteredMessage(
-                      'The "$page" screen is not available for this application.'),
-                  ideTheme: ideTheme,
-                  analyticsProvider: widget.analyticsProvider,
-                );
-              }
-              return _providedControllers(
-                child: DevToolsScaffold(
-                  embed: embed,
-                  ideTheme: ideTheme,
-                  initialPage: page,
-                  tabs: tabs,
-                  analyticsProvider: widget.analyticsProvider,
-                  actions: [
-                    // TODO(https://github.com/flutter/devtools/issues/1941)
-                    if (serviceManager.connectedApp.isFlutterAppNow) ...[
-                      HotReloadButton(),
-                      HotRestartButton(),
-                    ],
-                    OpenSettingsAction(),
-                    OpenAboutAction(),
-                  ],
-                ),
-              );
-            },
-          );
-        } else {
+  Widget _buildTabbedPage(
+      BuildContext context, String page, Map<String, String> params) {
+    final appServiceUrl = params['uri'];
```

Comment:
perhaps rename to `vmServiceUri` since that is what it is referred as throughout the code and documentation